slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/InsertTest.scala

```diff
@@ -230,4 +230,90 @@ class InsertTest extends AsyncTest[JdbcTestDB] {
       }
     } yield ()).withPinnedSession
   }
+
+  // TODO DerbyProfile and PostgresProfile seems to have a problem in this case. check #2206 and #2207
+  def testInsertOrUpdateWithPrimaryKeyOnly: DBIOAction[Unit, NoStream, Effect.All] = if (!tdb.confName.matches("(derby|postgres).*")) {
+    class T(tag: Tag) extends Table[Int](tag, "mytable") {
+      def id = column[Int]("id", O.PrimaryKey)
+      def * = id
+    }
+    case class V(id: Int, value: String)
+    class T2(tag: Tag) extends Table[V](tag, "mytable2") {
+      def id = column[Int]("id")
+      def value = column[String]("value", O.SqlType("varchar(20)"))
+      def * = (id, value) <> ((V.apply _).tupled, V.unapply)
+      def pk = primaryKey("pk", (id, value))
+    }
+    val ts1 = TableQuery[T]
+    val ts2 = TableQuery[T2]
+    (for {
+      _ <- ts1.schema.create
+      _ <- ts1 ++= Seq((1), (2))
+      _ <- ts1.insertOrUpdate((0)).map(_ shouldBe 1)
+      _ <- ts1.insertOrUpdate((1)).map(_ shouldBe 1)
+      _ <- ts2.schema.create
+      _ <- ts2 ++= Seq(V(1, "a"), V(2, "b"))
+      _ <- ts2.insertOrUpdate(V(0, "c")).map(_ shouldBe 1)
+      _ <- ts2.insertOrUpdate(V(1, "a")).map(_ shouldBe 1)
+      _ <- ts2.insertOrUpdate(V(2, "d")).map(_ shouldBe 1)
+    } yield ()).withPinnedSession
+  } else DBIO.seq()
+
+  def testInsertOrUpdateWithIntegrityError: DBIOAction[Unit, NoStream, Effect.All] = {
+    case class Book(id: Long, title: String, code: Option[String] = None)
+    class BooksTable(tag: Tag) extends Table[Book](tag, "books") {
+      def id = column[Long]("id", O.PrimaryKey)
+      def title = column[String]("title", O.SqlType("varchar(20)"))
+      def code = column[Option[String]]("code", O.SqlType("varchar(20)"))
```

Comment:
Why hardcode a O.SqlType when there's a O.Length for specifying string length in database-aware manner?

Reply:
Fixed in dbc35c8